### ISSUE TITLE
Extension: GraphQL

### ISSUE THREAD

**Issue by dhmlau** [strongloop/loopback-next#656]
# Description/Steps to reproduce

| Extension        | Descriptions           | Developers  |  LB4 Maintainer
| ------------- |:-------------:| -----:|-----:|
| GraphQL | Component to provide GraphQL Endpoint for Models. | @virkt25 , @mpicard, Help Welcome | Taranveer  (@virkt25) |
|       |      |    ||

# Additional information
This is originated from Call for Contribution on LB Extensions [issue](https://github.com/strongloop/loopback-next/issues/647).

**Comment by mpicard** [comment]
Looking forward to working on this! 😄

**Comment by dhmlau** [comment]
added @beeman.  Welcome aboard!

**Comment by beeman** [comment]
Thanks @dhmlau!

**Comment by raymondfeng** [comment]
FYI: Here are prior arts that integrate GraphQL with LoopBack 3.x:

- https://github.com/yahoohung/loopback-graphql-server
- https://github.com/Tallyb/loopback-graphql
- https://github.com/BlueEastCode/loopback-graphql-relay

**Comment by mpicard** [comment]
I'd like to look at https://github.com/apollographql/apollo-server as well as incorporating the https://github.com/apollographql/apollo-tracing interface for metrics! Would be cool to enable that.

**Comment by dhmlau** [comment]
@mpicard , now that you've mentioned apollo, here is another reference too. 
https://github.com/gramps-graphql/gramps-express

**Comment by mpicard** [comment]
@dhmlau very cool project! I wasn't aware of it. Thanks for sharing, the modular aspects of it look great and would be a great for lb!

**Comment by dhmlau** [comment]
Still needed!
I wonder if there's any news on this extension.  And would it be possible if we can use  https://github.com/yahoohung/loopback-graphql-server (just an example) as a reference for implementation?

**Comment by peanutenthusiast** [comment]
@dhmlau Any updates with this issue? My team is looking into using LB4 and GraphQL in production, so I'd be interested in providing a helping hand.

**Comment by dhmlau** [comment]
@peanutenthusiast, thanks for your interest!  We’re looking at an approach that would convert your REST APIs into GraphQL APIs.  Would that be something fit your need?  
It is a bit different from https://github.com/yahoohung/loopback-graphql-server which takes LB artifacts to convert into GraphQL APIs.  I can provide more details once we're (or close to) ready.

**Comment by peanutenthusiast** [comment]
@dhmlau Yes! This is exactly what my team is trying to do right now. Looking forward to further details.

**Comment by jjtang1** [comment]
@peanutenthusiast That is awesome to hear, would love you trying it out! Could you share what your team is trying to do just to help us understand the context? Thanks!

**Comment by peanutenthusiast** [comment]
@jjtang1 My team is interested in using LB4 as an orchestration layer API to aggregate needed data from different data sources (MongoDB, REST, SQLServer, etc.). Currently experimenting with a simple GraphQL LB4 server with an mssql connector, will update you and the team as we move further along the process.

**Comment by dhmlau** [comment]
@peanutenthusiast, we're proud to present to you the `oasgraph` module that allows you to create GraphQL APIs within the same LB application. 
http://v4.loopback.io/oasgraph.html

You can find more info on the integration here (docs will be on its way!): https://github.com/strongloop/loopback-next/issues/1775

**Comment by bajtos** [comment]
I updated issue description to better communicate the current status: we have an experimental version of GraphQL layer and looking for user feedback now.

You can learn more about the graphql extension in our docs: https://loopback.io/doc/en/lb4/GraphQL.html
